[Concurrency] `memberAccessHasSpecialPermissionInSwift5` should treat…

xedin opened the pull request:
… init accessors as stored properties

Init accessor properties cannot escape self and are only 
allowed to initialized and access a set of properties declared
in their `@storageRestrictions(...)` attribute.

Resolves: https://github.com/apple/swift/issues/70550

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[Concurrency] `memberAccessHasSpecialPermissionInSwift5` should treat init accessors as stored properties

Init accessor properties cannot escape self and are only allowed
to initialized and access a set of properties declared in their
`@storageRestrictions(...)` attribute.

Resolves: https://github.com/apple/swift/issues/70550

xedin commented:
@kavon I'd appreciate if you could help me come up with an example that should warn. I looked at the FlowIsolation path and it's ran after `assign_or_init` is expanded into individual accesses to underlying stored properties and/or setter/getter calls.

xedin commented:
Based on my read of FlowIsolation it doesn't matter what kind of access is performed inside, it tracks references.

kavon commented on test/Concurrency/flow_isolation.swift:
```diff
@@ -729,3 +729,28 @@ actor CheckDeinitFromActor {
+      self.degrees = degrees // Ok
+    }
+
+    init(radians: Double) {
+      self.radians = radians // Ok
+    }
```
Some suggestions for other things to test for:

1. What happens if you init `radians`, escape `self` from the initializer, and then try to access `radians` again?
2.  Same as above but if `radians` is a nonisolated property.
3. What happens if you try to escape `self` prior to `radians` being initialized?

I assume that for all of these, they _should_ work the same as a computed property (which is hopefully already tested).  Only (3) is one that is really to ensure FlowIsolation doesn't crash, since it should already get flagged as an error by DI prior.

kavon replied:
Otherwise if those are working, then this PR LGTM.

xedin replied:
@kavon I had to make some changes to FlowIsolation pass to recognize `setter` use of init accessor properties. That cannot be done earlier because only DI knows when to use init accessor and when to use a setter for a particular property.

xedin committed:
[SILOptimizer] FlowIsolation: Teach analysis to recognize use of init accessor properties

Init accessor properties should be handled specifically because
they do not reference underlying storage directly via `ref_element_addr`
instructions when 'self' is fully initialized.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

kavon commented on test/Concurrency/flow_isolation.swift:
```diff
@@ -734,3 +734,86 @@ actor CheckDeinitFromActor {
+
+      // degrees are initialized here via default value associated with radians
+
+      escapingSelf(self)
+
+      self.radians = 0
+      // expected-warning@-1 {{actor-isolated property 'radians' can not be mutated from a non-isolated context; this is an error in Swift 6}}
```
👍🏼

xedin commented:
@swift-ci please smoke test macOS platform
